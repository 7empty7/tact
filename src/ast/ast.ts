import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {
    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRef = {
    kind: 'type_ref',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction)[]
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public' } | { type: 'get' };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attribute: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementCall = {
    kind: 'statement_call',
    id: number,
    expression: ASTOpCall | ASTOpCallStatic,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: string[],
    expression: ASTExpression,
    ref: ASTRef
}

//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementCall | ASTSTatementAssign;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementCall | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_call' || src.kind === 'statement_assign';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}