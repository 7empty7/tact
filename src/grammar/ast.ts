import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {

    static merge(...refs: ASTRef[]) {
        if (refs.length === 0) {
            throw Error('Cannot merge 0 refs');
        }
        let r = refs[0].#interval;
        for (let i = 1; i < refs.length; i++) {
            r = r.coverageWith(r, refs[i].#interval);
        }
        return new ASTRef(r);
    }

    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTString = {
    kind: 'string',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

export type ASTLvalueRef = {
    kind: 'lvalue_ref',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRefSimple = {
    kind: 'type_ref_simple',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

export type ASTTypeRefMap = {
    kind: 'type_ref_map',
    id: number,
    key: string,
    value: string,
    ref: ASTRef
}

export type ASTTypeRef = ASTTypeRefSimple | ASTTypeRefMap;

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction | ASTProgramImport)[]
}

export type ASTProgramImport = {
    kind: 'program_import',
    id: number,
    path: ASTString,
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string,
    message: boolean,
    prefix: number | null,
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    init: bigint | boolean | null | undefined,
    as: string | null,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction | ASTInitFunction | ASTReceive)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public', ref: ASTRef } | { type: 'get', ref: ASTRef } | { type: 'mutates', ref: ASTRef } | { type: 'extends', ref: ASTRef };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTReceive = {
    kind: 'def_receive',
    id: number,
    selector: { kind: 'simple', arg: ASTArgument } | { kind: 'fallback' } | { kind: 'comment', comment: ASTString } | { kind: 'bounce', arg: ASTArgument },
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    attributes: ASTFunctionAttribute[],
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

export type ASTInitFunction = {
    kind: 'def_init_function',
    id: number,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementExpression = {
    kind: 'statement_expression',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: ASTLvalueRef[],
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTCondition = {
    kind: 'statement_condition',
    id: number,
    expression: ASTExpression,
    trueStatements: ASTStatement[],
    falseStatements: ASTStatement[];
    elseif: ASTCondition | null,
    ref: ASTRef,
}

export type ASTStatementWhile = {
    kind: 'statement_while'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementUntil = {
    kind: 'statement_until'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTStatementRepeat = {
    kind: 'statement_repeat'
    id: number,
    condition: ASTExpression,
    statements: ASTStatement[],
    ref: ASTRef
}


//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementExpression | ASTSTatementAssign | ASTCondition | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew | ASTNull | ASTLvalueRef;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementExpression | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef | ASTNull | ASTCondition | ASTInitFunction | ASTStatementWhile | ASTStatementUntil | ASTStatementRepeat | ASTReceive | ASTLvalueRef | ASTString | ASTProgramImport;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_expression' || src.kind === 'statement_assign' || src.kind === 'statement_condition' || src.kind === 'statement_while' || src.kind === 'statement_until' || src.kind === 'statement_repeat';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function __DANGER_resetNodeId() {
    nextId = 1;
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}