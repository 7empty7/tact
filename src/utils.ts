// // export type Interval = 
// export function thr