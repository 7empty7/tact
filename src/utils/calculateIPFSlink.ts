import { CID } from 'multiformats/cid';
import * as json from 'multiformats/codecs/json';
import { sha256 } from 'multiformats/hashes/sha2';

export async function calculateIPFSlink(data: Buffer) {
    const hash = await sha256.digest(data)
    const cid = CID.create(1, json.code, hash)
    return 'ipfs://' + cid.toString();
}
