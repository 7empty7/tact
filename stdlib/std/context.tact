struct Context {
    bounced: Bool;
    sender: Address;
    value: Int;
}

@name(__tact_context_get)
native context(): Context;