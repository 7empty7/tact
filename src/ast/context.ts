import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}