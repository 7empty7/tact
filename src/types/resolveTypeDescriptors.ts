import { ASTField, ASTFunction, ASTInitFunction, ASTNativeFunction, ASTTypeRef, throwError } from "../ast/ast";
import { CompilerContext, createContextStore } from "../ast/context";
import { FieldDescription, FunctionArgument, FunctionDescription, InitDescription, TypeDescription, TypeRef } from "./types";

let store = createContextStore<TypeDescription>();
let staticFunctionsStore = createContextStore<FunctionDescription>();

export function resolveTypeDescriptors(ctx: CompilerContext) {
    let types: { [key: string]: TypeDescription } = {};
    let staticFunctions: { [key: string]: FunctionDescription } = {};

    // Register types
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];
        if (types[a.name]) {
            throwError(`Type ${a.name} already exists`, a.ref);
        }
        if (a.kind === 'primitive') {
            types[a.name] = {
                kind: 'primitive',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        } else if (a.kind === 'def_contract') {
            types[a.name] = {
                kind: 'contract',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        } else if (a.kind === 'def_struct') {
            types[a.name] = {
                kind: 'struct',
                name: a.name,
                fields: [],
                functions: [],
                init: null
            };
        }
    }

    function resolveTypeRef(src: ASTTypeRef): TypeRef {
        if (!types[src.name]) {
            throwError('Type ' + src.name + ' not found', src.ref);
        }
        if (src.optional) {
            return {
                kind: 'optional',
                inner: {
                    kind: 'direct',
                    name: src.name
                }
            };
        } else {
            return {
                kind: 'direct',
                name: src.name
            }
        }
    }

    function resolveFunctionDescriptor(self: TypeDescription | null, a: ASTFunction | ASTNativeFunction): FunctionDescription {

        // Resolve return
        let returns: TypeRef | null = null;
        if (a.return) {
            returns = resolveTypeRef(a.return);
        }

        // Resolve args
        let args: FunctionArgument[] = [];
        for (let r of a.args) {
            args.push({
                name: r.name,
                type: resolveTypeRef(r.type),
                as: null
            });
        }

        // Register function
        return {
            name: a.name,
            self,
            args,
            returns,
            ast: a,
            isPublic: a.kind === 'def_function' ? a.attribute.some(a => a.type === 'public') : false,
            isGetter: a.kind === 'def_function' ? a.attribute.some(a => a.type === 'get') : false,
        };
    }

    function resolveInitFunction(a: ASTInitFunction): InitDescription {
        let args: FunctionArgument[] = [];
        for (let r of a.args) {
            args.push({
                name: r.name,
                type: resolveTypeRef(r.type),
                as: null
            });
        }
        return {
            args,
            ast: a
        }
    }

    // Resolve static functions
    for (let f in ctx.astFunctionStatic) {
        let a = ctx.astFunctionStatic[f];
        if (staticFunctions[a.name]) {
            throw Error('Function ' + a.name + ' already exists');
        }

        // Register function
        staticFunctions[a.name] = resolveFunctionDescriptor(null, a);
    }

    // Resolve fields
    function resolveField(src: ASTField, index: number): FieldDescription {
        return { name: src.name, type: resolveTypeRef(src.type), index, as: src.as };
    }
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];

        // Contract
        if (a.kind === 'def_contract') {
            for (const f of a.declarations) {
                if (f.kind !== 'def_field') {
                    continue;
                }
                if (types[a.name].fields.find((v) => v.name === f.name)) {
                    throw Error('Field ' + f.name + ' already exists');
                }
                types[a.name].fields.push(resolveField(f, types[a.name].fields.length));
            }
        }

        // Struct
        if (a.kind === 'def_struct') {
            for (let f of a.fields) {
                if (types[a.name].fields.find((v) => v.name === f.name)) {
                    throw Error('Field ' + f.name + ' already exists');
                }
                types[a.name].fields.push(resolveField(f, types[a.name].fields.length));
            }
        }
    }

    // Resolve contract functions
    for (let t in ctx.astTypes) {
        let a = ctx.astTypes[t];
        if (a.kind === 'def_contract') {
            let s = types[a.name];
            for (let d of a.declarations) {
                if (d.kind === 'def_function') {
                    s.functions.push(resolveFunctionDescriptor(s, d));
                }
                if (d.kind === 'def_init_function') {
                    if (s.init) {
                        throw Error('Init function already exists');
                    }
                    s.init = resolveInitFunction(d);
                }
            }
        }
    }

    // Register types in context
    for (let t in types) {
        ctx = store.set(ctx, t, types[t]);
    }
    for (let t in staticFunctions) {
        ctx = staticFunctionsStore.set(ctx, t, staticFunctions[t]);
    }

    return ctx;
}

export function getType(ctx: CompilerContext, name: string): TypeDescription {
    let r = store.get(ctx, name);
    if (!r) {
        throw Error('Type ' + name + ' not found');
    }
    return r;
}

export function getAllTypes(ctx: CompilerContext) {
    return store.all(ctx);
}

export function getStaticFunction(ctx: CompilerContext, name: string): FunctionDescription {
    let r = staticFunctionsStore.get(ctx, name);
    if (!r) {
        throw Error('Static function ' + name + ' not found');
    }
    return r;
}

export function getAllStaticFunctions(ctx: CompilerContext) {
    return staticFunctionsStore.all(ctx);
}