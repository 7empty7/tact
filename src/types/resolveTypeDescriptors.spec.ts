import { CompilerContext } from "../ast/context";
import { getAllStaticFunctions, getAllTypes, resolveTypeDescriptors } from "./resolveTypeDescriptors";
import fs from 'fs';
import { ASTRef, __DANGER_resetNodeId } from "../ast/ast";
import { loadCases } from "../utils/loadCases";

expect.addSnapshotSerializer({
    test: (src) => src instanceof ASTRef,
    print: (src) => `${(src as ASTRef).contents}`
});

describe('resolveTypeDescriptors', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    for (let r of loadCases(__dirname + "/test/")) {
        it('should resolve descriptors for ' + r.name, () => {
            let ctx = CompilerContext.fromSources([r.code]);
            ctx = resolveTypeDescriptors(ctx);
            expect(getAllTypes(ctx)).toMatchSnapshot();
            expect(getAllStaticFunctions(ctx)).toMatchSnapshot();
        });
    }
    for (let r of loadCases(__dirname + "/test-failed/")) {
        it('should fail descriptors for ' + r.name, () => {
            let ctx = CompilerContext.fromSources([r.code]);
            expect(() => resolveTypeDescriptors(ctx)).toThrowErrorMatchingSnapshot();
        });
    }
});