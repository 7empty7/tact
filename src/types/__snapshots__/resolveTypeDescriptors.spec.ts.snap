// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveTypeDescriptors should resolve descriptors for case-0.tact 1`] = `
{
  "Bool": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Bool",
  },
  "Int": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Int",
  },
}
`;

exports[`resolveTypeDescriptors should resolve descriptors for case-0.tact 2`] = `{}`;

exports[`resolveTypeDescriptors should resolve descriptors for case-1.tact 1`] = `
{
  "Bool": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Bool",
  },
  "Int": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Int",
  },
  "Struct1": {
    "fields": [
      {
        "index": 0,
        "name": "a",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Int",
        },
      },
      {
        "index": 1,
        "name": "b",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Bool",
        },
      },
    ],
    "functions": [],
    "kind": "struct",
    "name": "Struct1",
  },
  "Struct2": {
    "fields": [
      {
        "index": 0,
        "name": "a",
        "type": {
          "fields": [
            {
              "index": 0,
              "name": "a",
              "type": {
                "fields": [],
                "functions": [],
                "kind": "primitive",
                "name": "Int",
              },
            },
            {
              "index": 1,
              "name": "b",
              "type": {
                "fields": [],
                "functions": [],
                "kind": "primitive",
                "name": "Bool",
              },
            },
          ],
          "functions": [],
          "kind": "struct",
          "name": "Struct1",
        },
      },
    ],
    "functions": [],
    "kind": "struct",
    "name": "Struct2",
  },
}
`;

exports[`resolveTypeDescriptors should resolve descriptors for case-1.tact 2`] = `{}`;

exports[`resolveTypeDescriptors should resolve descriptors for case-2.tact 1`] = `
{
  "Bool": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Bool",
  },
  "Int": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Int",
  },
}
`;

exports[`resolveTypeDescriptors should resolve descriptors for case-2.tact 2`] = `
{
  "testFunc": {
    "args": [],
    "ast": {
      "args": [],
      "attribute": [],
      "id": 16,
      "kind": "def_function",
      "name": "testFunc",
      "ref": fun testFunc(): Int {
    return 0;
},
      "return": "Int",
      "statements": [
        {
          "expression": {
            "id": 14,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 15,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isPublic": false,
    "name": "testFunc",
    "returns": {
      "fields": [],
      "functions": [],
      "kind": "primitive",
      "name": "Int",
    },
    "self": null,
  },
  "testFunc2": {
    "args": [],
    "ast": {
      "args": [],
      "attribute": [],
      "id": 19,
      "kind": "def_function",
      "name": "testFunc2",
      "ref": fun testFunc2() {
    return 0;
},
      "return": null,
      "statements": [
        {
          "expression": {
            "id": 17,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 18,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isPublic": false,
    "name": "testFunc2",
    "returns": null,
    "self": null,
  },
  "testFunc3": {
    "args": [
      {
        "name": "a",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Int",
        },
      },
      {
        "name": "b",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Bool",
        },
      },
    ],
    "ast": {
      "args": [
        {
          "id": 20,
          "kind": "def_argument",
          "name": "a",
          "ref": a: Int,
          "type": "Int",
        },
        {
          "id": 21,
          "kind": "def_argument",
          "name": "b",
          "ref": b: Bool,
          "type": "Bool",
        },
      ],
      "attribute": [],
      "id": 24,
      "kind": "def_function",
      "name": "testFunc3",
      "ref": fun testFunc3(a: Int, b: Bool) {
    return 0;
},
      "return": null,
      "statements": [
        {
          "expression": {
            "id": 22,
            "kind": "number",
            "ref": 0,
            "value": 0n,
          },
          "id": 23,
          "kind": "statement_return",
          "ref": return 0;,
        },
      ],
    },
    "isGetter": false,
    "isPublic": false,
    "name": "testFunc3",
    "returns": null,
    "self": null,
  },
}
`;

exports[`resolveTypeDescriptors should resolve descriptors for case-3.tact 1`] = `
{
  "Bool": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Bool",
  },
  "HelloWorld": {
    "fields": [
      {
        "index": 0,
        "name": "a",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Int",
        },
      },
      {
        "index": 1,
        "name": "b",
        "type": {
          "fields": [],
          "functions": [],
          "kind": "primitive",
          "name": "Bool",
        },
      },
    ],
    "functions": [
      {
        "args": [],
        "ast": {
          "args": [],
          "attribute": [],
          "id": 30,
          "kind": "def_function",
          "name": "hello",
          "ref": fun hello() {
        
    },
          "return": null,
          "statements": [],
        },
        "isGetter": false,
        "isPublic": false,
        "name": "hello",
        "returns": null,
        "self": [Circular],
      },
      {
        "args": [],
        "ast": {
          "args": [],
          "attribute": [],
          "id": 31,
          "kind": "def_function",
          "name": "hello2",
          "ref": fun hello2() {
        
    },
          "return": null,
          "statements": [],
        },
        "isGetter": false,
        "isPublic": false,
        "name": "hello2",
        "returns": null,
        "self": [Circular],
      },
    ],
    "kind": "contract",
    "name": "HelloWorld",
  },
  "Int": {
    "fields": [],
    "functions": [],
    "kind": "primitive",
    "name": "Int",
  },
}
`;

exports[`resolveTypeDescriptors should resolve descriptors for case-3.tact 2`] = `{}`;
